packaging/installer/functions.sh

```diff
@@ -641,13 +641,20 @@ issystemd() {
 }
 
 get_systemd_service_dir() {
-  if [ -w "/lib/systemd/system" ]; then
-    echo "/lib/systemd/system"
-  elif [ -w "/usr/lib/systemd/system" ]; then
-    echo "/usr/lib/systemd/system"
-  elif [ -w "/etc/systemd/system" ]; then
-    echo "/etc/systemd/system"
-  fi
+  unit_paths="$(systemctl show -p UnitPath --value | tr ' ' '\n')"
+
+  # The strange `sed` expression here reverses line order (systemd lists
+  # unit paths in descending priority, we want to check the lib paths in
+  # ascending priority)
+  lib_paths="$(echo "${unit_paths}" | grep -vE '^/\(run|etc\)' | sed '1!x;H;1h;$!d;g')"
+  etc_paths="$(echo "${unit_paths}" | grep -E '^/etc' | grep -vE '(attached|control)$')"
+
+  for path in ${lib_paths} ${etc_paths}; do
+    if [ -d "${path}" ] && [ -w "${path}" ]; then
+      echo "${path}"
+      return 0
+    fi
+  done
```

Comment:
1. No error handling if `systemctl show` fails. For instance, old systemd versions don't have `--value`.
2. No explicit failure return code if no systemd service directory is found.
3. grep pattern `^/\(run|etc\)` is incorrect. Did you want `^/\(run|etc)`?
4. the `sed` expression is cryptic, consider using [tac](https://man7.org/linux/man-pages/man1/tac.1.html).

Reply:
1. I missed this (I keep forgetting we have to support ancient systemd versions, and thus forgot to check exactly how recent the option was added).
2. This is consistent with how the function worked before the change in this particular code, and is also what the calling code expects (it checks whether it got a string or not, not what the return code was). I’m being very intentional in trying to keep the changes minimal here to limit the possibility of introducing bugs.
3. Not sure how I missed this.
4. `sed` is portable, `tac` is not. I understand that the `sed` expression is non-obvious, but that’s the whole reason I have a comment there to explain it.

Reply:
> We expect this to work on Linux systems with systemd. Do you think there are any where this will be a problem?

This is a script that is expected to run on all UNIX-like platforms we support, and arbitrarily deciding ‘this code path will only ever be run under these circumstances’ is detrimental to maintaining that support.

For the record, it is in fact possible to use systemd without GNU coreutils, and thus without `tac`.

> This is more readable awk alternative

Point taken on this one. I tend to reach for sed for more complex text manipulations like this because it’s not unusual for AWK to be able to do what is needed, but in this case I just never looked at AWK.